### PR TITLE
Support multiple outbound topics for publishing blockchain events

### PR DESCRIPTION
**Pull request by abhakish**
With this change, based on configurations the block/chaincode events can now be sent to multiple topics.

### CODE REVIEW

File: src/main/java/hlf/java/rest/client/service/impl/EventPublishServiceImpl.java [hunk 1]
```diff
@@ -82,18 +84,51 @@ public void onFailure(Throwable ex) {
...
+          chaincodeName,
+          fabricTxId,
+          eventName,
+          channelName,
+          messageKey);
+      return;
+    }
+
+    for (String topic : optionalChaincodeDetails.get().getListenerTopics()) {
+      sendMessage(topic, payload, chaincodeName, fabricTxId, eventName, channelName, messageKey);
```
Comment (nithin-pankaj) [inline src/main/java/hlf/java/rest/client/service/impl/EventPublishServiceImpl.java:117]: Does all these topics rely on the same Kafka Template? What if they belong to different outbound brokers?

**Review by nithin-pankaj — CHANGES_REQUESTED** [review]

File: src/main/java/hlf/java/rest/client/service/impl/EventPublishServiceImpl.java (hunk 1 above)
Comment (abhakish) [reply src/main/java/hlf/java/rest/client/service/impl/EventPublishServiceImpl.java:117]: All topics rely on the same kafka template. We can provide multiple brokers in `brokerHost` field in application.yml in comma separated manner

Comment (abhakish) [reply src/main/java/hlf/java/rest/client/service/impl/EventPublishServiceImpl.java:120]: updated code to create multiple producers and using `RoutingKafkaTemplate` to dispatch message based on topic name configured while creating producer factory

File: src/main/java/hlf/java/rest/client/service/impl/EventPublishServiceImpl.java [hunk 2]
```diff
@@ -82,18 +87,51 @@ public void onFailure(Throwable ex) {
...
+        || CollectionUtils.isEmpty(optionalChaincodeDetails.get().getListenerTopics())) {
+      sendMessage(
+          kafkaProperties.getEventListeners().get(0).getTopic(),
+          payload,
+          chaincodeName,
+          fabricTxId,
+          eventName,
+          channelName,
+          messageKey);
+      return;
```
Comment (nithin-pankaj) [inline src/main/java/hlf/java/rest/client/service/impl/EventPublishServiceImpl.java:116]: Why should we publish the chaincode details are not present in the Events config?

File: src/main/resources/application.template [hunk 3]
```diff
@@ -47,14 +49,14 @@ kafka:
       ssl-enabled: boolean
       security-protocol: <Only supports SSL>
       ssl-keystore-base64: <if ssl-enabled is true, provide the Base64 encoded value of keystore file>
-      ssl-truststore-base64: <if ssl-enabled is true, provide the Base64 encoded value of Trsutstore file>
+      ssl-truststore-base64: <if ssl-enabled is true, provide the Base64 encoded value of Truststore file>
       offsetResetPolicy: <possible values are earliest / latest> if not provided, default will be latest.
-  event-listener:
+  event-listeners: <Note : This is a list and Multiple event listeners can be configured>
     brokerHost: <Comma separated list of boostrap servers>
```
[the remark this answers is not in the document]
Comment (abhakish) [reply src/main/resources/application.template]: updated

File: src/main/java/hlf/java/rest/client/service/impl/EventPublishServiceImpl.java (hunk 2 above)
Comment (abhakish) [reply src/main/java/hlf/java/rest/client/service/impl/EventPublishServiceImpl.java]: corrected

File: src/main/java/hlf/java/rest/client/service/impl/EventPublishServiceImpl.java [hunk 4]
```diff
@@ -4,27 +4,29 @@
...
 
   @Autowired private KafkaProperties kafkaProperties;
 
   @Autowired private FabricProperties fabricProperties;
 
-  @Autowired private KafkaTemplate<String, String> kafkaTemplate;
+  @Autowired private RoutingKafkaTemplate routingKafkaTemplate;
 
   @Override
   public boolean sendMessage(String msg, String fabricTxId, String eventName, String channelName) {
```
Comment (nithin-pankaj) [inline src/main/java/hlf/java/rest/client/service/impl/EventPublishServiceImpl.java:32]: Seems like this method is unused, could you also remove this from the service?

**Review by nithin-pankaj — APPROVED** [review]

**Pull request merged**